### PR TITLE
Performance improvements using iterator aggregates

### PR DESCRIPTION
**Pull request by drupol**
This PR:

- [x] Replaces `IterableIterator` with `IterableIteratorAggregate` (4 occurences)
- [x] Changes the signature of a couple of operations. 
- [x] Update all the annotations where an `Iterator` is needed while now an `iterable` is enough.
- [x] Use `yield from` - align everything
- [x] Get rid of `ClosureIterator` and replace it with `ClosureIteratorAggregate`

Based on the benchmarks I did in `loophp/iterators`, it seems to be definitely, and it make sense.
`IteratorAggregates` are usually faster because the state of the `Iterator` is not built in _userland_ but directly handled by PHP internaly.

```
$ ./vendor/bin/phpbench run --profile=local --group=local --report=aggregate --stop-on-error tests/benchmarks/IterableIteratorAggregateBench.php tests/benchmarks/IterableIteratorBench.php
PHPBench (1.2.3) running benchmarks...
with configuration file: /home/pol/dev/git/iterators/phpbench.json
with PHP version 8.1.1, xdebug ❌, opcache ✔

\benchmarks\loophp\iterators\IterableIteratorAggregateBench

    bench # IterableIteratorAggregate.......I0 - Mo1.963ms (±0.00%)
    bench # IterableIteratorAggregate.......R1 I7 - Mo0.937ms (±1.97%)
    bench # IterableIteratorAggregate.......R1 I25 - Mo0.927ms (±1.60%)

\benchmarks\loophp\iterators\IterableIteratorBench

    bench # IterableIterator................R1 I0 - Mo2.188ms (±0.00%)
    bench # IterableIterator................R1 I6 - Mo1.765ms (±1.80%)
    bench # IterableIterator................R4 I49 - Mo1.741ms (±2.25%)

Subjects: 2, Assertions: 0, Failures: 0, Errors: 0
+--------------------------------+---------+---------------------------+------+-----+----------+---------+--------+
| benchmark                      | subject | set                       | revs | its | mem_peak | mode    | rstdev |
+--------------------------------+---------+---------------------------+------+-----+----------+---------+--------+
| IterableIteratorAggregateBench | bench   | IterableIteratorAggregate | 10   | 1   | 3.695mb  | 1.963ms | ±0.00% |
| IterableIteratorAggregateBench | bench   | IterableIteratorAggregate | 10   | 10  | 3.695mb  | 0.937ms | ±1.97% |
| IterableIteratorAggregateBench | bench   | IterableIteratorAggregate | 10   | 50  | 3.695mb  | 0.927ms | ±1.60% |
| IterableIteratorBench          | bench   | IterableIterator          | 10   | 1   | 3.695mb  | 2.188ms | ±0.00% |
| IterableIteratorBench          | bench   | IterableIterator          | 10   | 10  | 3.695mb  | 1.765ms | ±1.80% |
| IterableIteratorBench          | bench   | IterableIterator          | 10   | 50  | 3.695mb  | 1.741ms | ±2.25% |
+--------------------------------+---------+---------------------------+------+-----+----------+---------+--------+
```

@AlexandruGG WDYT of this? Are you ok with this?

### CODE REVIEW

**Comment by AlexandruGG** [comment]
Looks good at a first glance, let me know when it's ready for review and I'll go through it more in-depth 👍

**Comment by drupol** [comment]
I think it's done.

This is something I should have done a long time ago.

**Comment by drupol** [comment]
Made a wrong commit, will fix this during the afternoon !

File: composer.json [hunk 1]
```diff
@@ -32,7 +32,7 @@
     ],
     "require": {
         "php": ">= 7.4",
-        "loophp/iterators": "^1.5.13"
+        "loophp/iterators": "^1.5.14"
```
Comment (AlexandruGG) [inline composer.json:35]: do we need to mention the minor version specifically or can we have this as just:
`"loophp/iterators": "^1.5"`

File: src/Contract/Operation/Duplicateable.php [hunk 2]
```diff
@@ -32,10 +32,8 @@ interface Duplicateable
      *
      * @see https://loophp-collection.readthedocs.io/en/stable/pages/api.html#duplicate
      *
-     * @template U
-     *
-     * @param null|callable(U): (Closure(U): bool) $comparatorCallback
-     * @param null|callable(T, TKey): U $accessorCallback
+     * @param (callable(mixed): (Closure(mixed): bool)) $comparatorCallback
+     * @param callable(T, TKey): mixed $accessorCallback
```
Comment (AlexandruGG) [inline src/Contract/Operation/Duplicateable.php:36]: these seem better done in a separate PR as they don't seem related to the iterators work.

Also, again, needs to be in line with the operation itself; but in the operation the template `U` is still there: https://github.com/loophp/collection/pull/233/files#diff-91532943eb0b11a568f0217fd47cb706a322494a906f5f2884ee76a9d71e60d1R30

File: composer.json (hunk 1 above)
Comment (drupol) [reply composer.json:35]: Yeah we could do that, it's the same.

File: src/Contract/Operation/Duplicateable.php (hunk 2 above)
Comment (drupol) [reply src/Contract/Operation/Duplicateable.php:36]: Ok !

File: src/Contract/Operation/Associateable.php [hunk 3]
```diff
@@ -23,13 +22,10 @@ interface Associateable
      *
      * @see https://loophp-collection.readthedocs.io/en/stable/pages/api.html#associate
      *
-     * @template NewT
-     * @template NewTKey
```
Comment (AlexandruGG) [inline src/Contract/Operation/Associateable.php:27]: similar to the other one, this template param change can be done in a separate PR

File: src/Contract/Operation/Duplicateable.php [hunk 4]
```diff
@@ -30,8 +30,6 @@ interface Duplicateable
      * of the current iterated value and then return the value to compare.
      * This is useful when you want to compare objects.
      *
-     * @see https://loophp-collection.readthedocs.io/en/stable/pages/api.html#duplicate
```
Comment (AlexandruGG) [inline src/Contract/Operation/Duplicateable.php:33]: this should not be removed

File: src/Contract/Operation/Associateable.php (hunk 3 above)
Comment (drupol) [reply src/Contract/Operation/Associateable.php]: Yes, fixed !

File: src/Contract/Operation/Duplicateable.php (hunk 4 above)
Comment (drupol) [reply src/Contract/Operation/Duplicateable.php]: Oops... nice catch!

Comment (drupol) [reply src/Contract/Operation/Duplicateable.php]: Fixed.

File: src/Collection.php [hunk 5]
```diff
@@ -156,40 +155,32 @@
...
-    private iterable $parameters;
+    private IteratorAggregate $innerIterator;
 
     /**
-     * @var callable(mixed ...$parameters): iterable<TKey, T>
-     */
-    private $source;
-
-    /**
-     * @psalm-external-mutation-free
```
Comment (AlexandruGG) [inline src/Collection.php:169]: is removing this intentional?

File: src/Collection.php [hunk 6]
```diff
@@ -503,62 +480,58 @@ public static function fromGenerator(Generator $generator): self
...
      * @return self<int, string>
      */
     public static function fromResource($resource): self
     {
-        return new self(static fn (): IteratorAggregate => new ResourceIteratorAggregate($resource));
+        return new self(static fn (): Generator => yield from new ResourceIteratorAggregate($resource));
     }
 
     /**
-     * @pure
```
Comment (AlexandruGG) [inline src/Collection.php:522]: I just noticed some of these `@pure` annotations are being removed, but not everywhere. I think just in `src/Collection.php`?

How come this is needed?

File: src/Collection.php (hunk 5 above)
Comment (drupol) [reply src/Collection.php:169]: Yeah, I think we should even get rid of them everywhere. psalm doesn't handle them correctly anyway. I really wanted to avoid updating the baselines too much.

File: src/Collection.php (hunk 6 above)
Comment (drupol) [reply src/Collection.php:522]: Same comment as the previous one. We are not the only ones... https://github.com/marcosh/lamphpda/blob/master/adr/2022-01-25-pure-callable.md

Comment (AlexandruGG) [reply src/Collection.php:522]: > Same comment as the previous one. We are not the only ones... https://github.com/marcosh/lamphpda/blob/master/adr/2022-01-25-pure-callable.md

I get this one, but in this project we didn't go that far to use `pure-callable` types 😬 .

I think the errors you saw in this project might be related to the iterators project not having things marked as `@pure`.

In any case it might be too much of a hassle to maintain this kind of stuff in PHP, in which case I think it's fine to remove them but we can do it for the whole codebase. You can do it in a separate PR though

**Review by AlexandruGG — APPROVED** [review]
